File: scripts/integrations/troubleshoot.py
#!/usr/bin/env python3
"""
troubleshoot.py

PURPOSE: Diagnose common issues with Prismatic integration development

USAGE: python troubleshoot.py [project-directory]

PARAMETERS:
  project-directory - (Optional) Path to integration project

EXIT CODES:
  Returns number of issues found (0 = all checks passed)

"""

import os
import subprocess
import sys

SHARED_DIR = os.path.join(os.path.dirname(__file__), '..', 'shared')
sys.path.insert(0, SHARED_DIR)

from graphql import ensure_authenticated, GraphQLError
from prism_retry import is_auth_error, is_network_error, run_prism_query


def check_prism_cli():
    """Check if Prism CLI is installed."""
    try:
        ensure_authenticated()
    except GraphQLError:
        pass  # check_network_and_auth() handles the diagnostic

    try:
        result = subprocess.run(
            ["prism", "--version"],
            capture_output=True,
            text=True,
            env=os.environ,
            timeout=5,
        )
        if result.returncode == 0:
            version = result.stdout.strip()
            return True, f"Prism CLI: Installed ({version})"
        else:
            return False, "Prism CLI: Installed but not working correctly"
    except FileNotFoundError:
        return (
            False,
            "Prism CLI: Not installed (run setup_prerequisites.py first)",
        )
    except Exception as e:
        return False, f"Prism CLI: Error checking ({str(e)})"


def check_network_and_auth():
    """Check network access and authentication to Prismatic (single call)."""
    try:
        result = run_prism_query(["prism", "me"], timeout=10)

        if result.returncode == 0:
            lines = result.stdout.strip().split("\n")
            if lines and lines[0]:
                return True, f"Network & Auth: Connected as {lines[0]}"
            return True, "Network & Auth: Connected and authenticated"

        error_text = (result.stderr or "") + " " + (result.stdout or "")

        if is_network_error(error_text):
            return False, "Network: Cannot reach *.prismatic.io (check firewall/proxy)"

        if is_auth_error(error_text):
            return False, "Auth: Not authenticated (run setup_prerequisites.py)"

        return False, "Network & Auth: Connection failed"

    except subprocess.TimeoutExpired:
        return False, "Network: Connection timeout to Prismatic"
    except FileNotFoundError:
        return None, "Network & Auth: Cannot test (prism CLI not installed)"
    except Exception as e:
        return False, f"Network & Auth: Error ({str(e)})"


def check_node():
    """Check Node.js version."""
    try:
        result = subprocess.run(
            ["node", "--version"], capture_output=True, text=True, timeout=5
        )
        if result.returncode == 0:
            version = result.stdout.strip()
            major_version = int(version.lstrip("v").split(".")[0])

            if major_version >= 18:
                return True, f"Node.js: {version} (compatible)"
            else:
                return False, f"Node.js: {version} (requires v18+ for Prismatic)"
        else:
            return False, "Node.js: Installed but not working"
    except FileNotFoundError:
        return False, "Node.js: Not installed (download from nodejs.org)"
    except Exception as e:
        return False, f"Node.js: Error ({str(e)})"


def check_project(project_dir):
    """Check project structure and compilation."""
    if not project_dir:
        return None, "Project: Not specified"

    if not os.path.isdir(project_dir):
        return False, f"Project: Directory not found ({project_dir})"

    package_json = os.path.join(project_dir, "package.json")
    if not os.path.isfile(package_json):
        return False, "Project: Missing package.json"

    src_dir = os.path.join(project_dir, "src")
    if not os.path.isdir(src_dir):
        return False, "Project: Missing src/ directory"

    node_modules = os.path.join(project_dir, "node_modules")
    if not os.path.isdir(node_modules):
        return False, "Project: Dependencies not installed (run: npm install)"

    dist_dir = os.path.join(project_dir, "dist")
    if not os.path.isdir(dist_dir):
        return None, "Project: Not built yet (run: npm run build)"

    return True, "Project: Structure valid and built"


def troubleshoot(project_dir=None):
    """Run all diagnostic checks."""
    print("🔍 Troubleshooting Prismatic Integration Development Environment")
    print("")

    checks = [
        ("Prism CLI", check_prism_cli),
        ("Node.js", check_node),
        ("Network & Auth", check_network_and_auth),
    ]

    if project_dir:
        checks.append(("Project", lambda: check_project(project_dir)))

    issues = 0
    warnings = 0

    for name, check_func in checks:
        try:
            passed, message = check_func()
            if passed:
                print(f"✅ {message}")
            elif passed is None:
                print(f"⚠️  {message}")
                warnings += 1
            else:
                print(f"❌ {message}")
                issues += 1
        except Exception as e:
            print(f"❌ {name}: Unexpected error - {e}")
            issues += 1

    print("")
    print("=" * 60)

    if issues == 0 and warnings == 0:
        print("✅ All checks passed! Your environment is ready.")
        print("")
        print("📋 Next steps:")
        if project_dir:
            print("  - Continue working on your integration")
            print("  - Run: python build_integration.py <project>")
            print("  - Deploy: python deploy_integration.py <project>")
        else:
            print("  - Run setup: python setup_prerequisites.py <name> <token>")
            print("  - Search components: python search_components.py <term>")
    elif issues == 0:
        print(f"⚠️  All critical checks passed, but {warnings} warning(s) found.")
        print("")
        print("Review the warnings above and address if needed.")
    else:
        print(f"❌ Found {issues} issue(s) and {warnings} warning(s)")
        print("")
        print("📋 Common fixes:")
        print("  - Run setup: python setup_prerequisites.py <name> <token> [url]")
        print("  - Check network access to *.prismatic.io")

    print("=" * 60)

    return issues


def main():
    project_dir = sys.argv[1] if len(sys.argv) > 1 else None
    return troubleshoot(project_dir)


if __name__ == "__main__":
    sys.exit(main())

File: plugin/scripts/integrations/troubleshoot.ts
#!/usr/bin/env node
/**
 * troubleshoot.ts
 *
 * PURPOSE: Diagnose common issues with Prismatic integration development
 *
 * USAGE: node troubleshoot.ts [project-directory]
 *
 * EXIT CODES:
 *   Returns number of issues found (0 = all checks passed)
 */

import { existsSync, statSync } from "node:fs";
import { join } from "node:path";
import { exec } from "../../vendor/tinyexec/main.mjs";
import { ensureAuthenticated } from "../shared/graphql.ts";
import { isAuthError, isNetworkError, runPrismQuery } from "../shared/prism-retry.ts";

const checkPrismCli = async (): Promise<[boolean | null, string]> => {
  try {
    await ensureAuthenticated();
  } catch {
    // check_network_and_auth handles the diagnostic
  }

  try {
    const result = await exec("prism", ["--version"], { timeout: 5000 });
    if (result.exitCode === 0) {
      const version = result.stdout.trim();
      return [true, `Prism CLI: Installed (${version})`];
    }
    return [false, "Prism CLI: Installed but not working correctly"];
  } catch (e) {
    if (e instanceof Error && e.message.includes("ENOENT")) {
      return [false, "Prism CLI: Not installed (run setup prerequisites first)"];
    }
    return [false, `Prism CLI: Error checking (${e})`];
  }
};

const checkNetworkAndAuth = async (): Promise<[boolean | null, string]> => {
  try {
    const result = await runPrismQuery(["prism", "me"], 10);

    if (result.returncode === 0) {
      const lines = result.stdout.trim().split("\n");
      if (lines[0]) {
        return [true, `Network & Auth: Connected as ${lines[0]}`];
      }
      return [true, "Network & Auth: Connected and authenticated"];
    }

    const errorText = `${result.stderr || ""} ${result.stdout || ""}`;

    if (isNetworkError(errorText)) {
      return [false, "Network: Cannot reach *.prismatic.io (check firewall/proxy)"];
    }

    if (isAuthError(errorText)) {
      return [false, "Auth: Not authenticated (run setup prerequisites)"];
    }

    return [false, "Network & Auth: Connection failed"];
  } catch (e) {
    if (e instanceof Error) {
      if (e.message.includes("TIMEOUT")) {
        return [false, "Network: Connection timeout to Prismatic"];
      }
      if (e.message.includes("ENOENT")) {
        return [null, "Network & Auth: Cannot test (prism CLI not installed)"];
      }
    }
    return [false, `Network & Auth: Error (${e})`];
  }
};

const checkNode = async (): Promise<[boolean | null, string]> => {
  try {
    const result = await exec(process.execPath, ["--version"], { timeout: 5000 });
    if (result.exitCode === 0) {
      const version = result.stdout.trim();
      const [majorVersion, minorVersion] = version
        .replace(/^v/, "")
        .split(".")
        .map((part) => Number.parseInt(part, 10));

      if (majorVersion > 22 || (majorVersion === 22 && minorVersion >= 18)) {
        return [true, `Node.js: ${version} (compatible)`];
      }
      return [false, `Node.js: ${version} (requires v22.18+ for native TypeScript)`];
    }
    return [false, "Node.js: Installed but not working"];
  } catch (e) {
    if (e instanceof Error && e.message.includes("ENOENT")) {
      return [false, "Node.js: Not installed (download from nodejs.org)"];
    }
    return [false, `Node.js: Error (${e})`];
  }
};

function checkProject(projectDir: string | null): [boolean | null, string] {
  if (!projectDir) return [null, "Project: Not specified"];

  try {
    if (!statSync(projectDir).isDirectory()) {
      return [false, `Project: Directory not found (${projectDir})`];
    }
  } catch {
    return [false, `Project: Directory not found (${projectDir})`];
  }

  if (!existsSync(join(projectDir, "package.json"))) {
    return [false, "Project: Missing package.json"];
  }

  try {
    if (!statSync(join(projectDir, "src")).isDirectory()) {
      return [false, "Project: Missing src/ directory"];
    }
  } catch {
    return [false, "Project: Missing src/ directory"];
  }

  try {
    if (!statSync(join(projectDir, "node_modules")).isDirectory()) {
      return [false, "Project: Dependencies not installed (run: npm install)"];
    }
  } catch {
    return [false, "Project: Dependencies not installed (run: npm install)"];
  }

  try {
    if (!statSync(join(projectDir, "dist")).isDirectory()) {
      return [null, "Project: Not built yet (run: npm run build)"];
    }
  } catch {
    return [null, "Project: Not built yet (run: npm run build)"];
  }

  return [true, "Project: Structure valid and built"];
}

const troubleshoot = async (projectDir: string | null): Promise<number> => {
  console.log("Troubleshooting Prismatic Integration Development Environment");
  console.log("");

  const checks: Array<[string, () => Promise<[boolean | null, string]>]> = [
    ["Prism CLI", checkPrismCli],
    ["Node.js", checkNode],
    ["Network & Auth", checkNetworkAndAuth],
  ];

  if (projectDir) {
    checks.push(["Project", async () => checkProject(projectDir)]);
  }

  let issues = 0;
  let warnings = 0;

  for (const [name, checkFn] of checks) {
    try {
      const [passed, message] = await checkFn();
      if (passed === true) {
        console.log(`  OK: ${message}`);
      } else if (passed === null) {
        console.log(`  WARN: ${message}`);
        warnings++;
      } else {
        console.log(`  FAIL: ${message}`);
        issues++;
      }
    } catch (e) {
      console.log(`  FAIL: ${name}: Unexpected error - ${e}`);
      issues++;
    }
  }

  console.log("");
  console.log("=".repeat(60));

  if (issues === 0 && warnings === 0) {
    console.log("All checks passed! Your environment is ready.");
    console.log("");
    console.log("Next steps:");
    if (projectDir) {
      console.log("  - Continue working on your integration");
      console.log("  - Build: node build-integration.ts <project>");
      console.log("  - Deploy: node deploy-integration.ts <project>");
    } else {
      console.log("  - Run setup prerequisites");
      console.log("  - Search components: node find-components.ts <term>");
    }
  } else if (issues === 0) {
    console.log(`All critical checks passed, but ${warnings} warning(s) found.`);
    console.log("");
    console.log("Review the warnings above and address if needed.");
  } else {
    console.log(`Found ${issues} issue(s) and ${warnings} warning(s)`);
    console.log("");
    console.log("Common fixes:");
    console.log("  - Run setup prerequisites");
    console.log("  - Check network access to *.prismatic.io");
  }

  console.log("=".repeat(60));

  return issues;
};

const main = async (): Promise<number> => {
  const projectDir = process.argv[2] ?? null;
  return troubleshoot(projectDir);
};

process.exit(await main());
